src/generators/gencpp.ml

```diff
@@ -3088,7 +3088,8 @@ let retype_expression ctx request_type function_args function_type expression_tr
                baseCpp.cppexpr, baseCpp.cpptype (* nothing to do *)
             else (match return_type with
             | TCppNativePointer(klass) -> CppCastNative(baseCpp), return_type
-            | TCppVoid -> baseCpp.cppexpr, TCppVoid
+            | TCppVoid ->
+				CppTCast(baseCpp, cpp_type_of expr.etype), return_type
```

Comment:
@hughsando: Please double-check this! The problem here is that all block-level expressions (except the last one of course) are typed against `TCppVoid`, so any safe-cast would be ignored here. My change generates the cast, but pipes `return_type` through so it's still considered to be `Void`.

Reply:
This should solve the problem I think.  Looks like this code should be using the supplied type, rather than the return type, but I'm not too keen to mess with it.  Maybe they are always the same unless it's Void.